src/App.js

```diff
@@ -59,42 +61,61 @@ class App extends Component {
         <Route exact path="/login" component={Login} />
         <Route exact path="/signup" component={Signup} />
         <Route exact path="/projects/new" component={NewProject} />
+      </>
+    );
+  };
+
+  constructComponent = (componentName) => {
+    let route = "/projects/p/:hash/";
```

Comment:
I have had to change from using :hash to :unit_hash as hash is a reserved word in Ruby.

Reply:
for the url this doesnt matter, when sending to the api the header can be unit_hash